client/src/components/App.jsx

```diff
@@ -130,16 +131,39 @@ class App extends React.Component {
   }
 
   setBadge(productIdx, badge) {
+    
     const { products } = this.state;
-    const newProduct = {
-      ...products[productIdx],
-      badge,
-    };
-    const newProducts = [...products];
-    newProducts[productIdx] = newProduct;
-    this.setState({
-      products: newProducts,
-    });
+    const productImage = products[productIdx].image;
+    const newBadge = new Promise ((resolve, reject) => {this.checkBadge(resolve, badge, productImage)});
```

Comment:
Does this have to be a Promise? I don't think you're doing anything async if I understand it right. Would be nice to simplify it if you can. And maybe checkBadge does not need to be a separate function, could it just be part of setBadge? Or would it be too complicated to read? What are your thoughts?